File: original/j3d_with_likwid/j3d.c
// Jacobi 3D skeleton program
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "timing.h"
#include <likwid.h>

void dummy(double*, double*);

const double oos=1./6.;

void jacobi_line(double* d, const double* s,
                 const double* top, const double* bottom,
                 const double* front, const double* back, int n) {
                 int i,start=1;
//#pragma vector nontemporal
                 for(i=1; i<n-1; ++i) {
                     d[i] = oos*(s[i-1]+s[i+1]+top[i]+bottom[i]+front[i]+back[i]);
                 }
}

int main(int argc, char** argv) {

  double wct_start,wct_end,cput_start,cput_end,runtime,r;
  int iter,size,i,j,k,n,t0,t1,t;
  int ofs;
  double* phi[2];

  LIKWID_MARKER_INIT;
#pragma omp parallel
{
  LIKWID_MARKER_THREADINIT;
}


  if(argc!=2) {
    printf("Usage: %s <size>\n",argv[0]);
    exit(1);
  }

  size = atoi(argv[1]);
  phi[0] = malloc((size_t)size*size*size*sizeof(double));
  phi[1] = malloc((size_t)size*size*size*sizeof(double));

  t0=0; t1=1;

  /* initialize w/ random numbers */
#pragma omp parallel for private(j,ofs)
  for(i=0; i<size; ++i) {
    for(j=0; j<size; ++j) {
      ofs = i*size*size + j*size;
      for(k=0; k<size; ++k) {
	phi[t1][ofs+k] = phi[t0][ofs+k] = rand()/(double)RAND_MAX;
      }
    }
  }	  

  iter=1;
  runtime=0.0;
  while(runtime<.5) {

  // time measurement
  timing(&wct_start, &cput_start);
  for(n=0; n<iter; n++) {
#pragma omp parallel
{
    LIKWID_MARKER_START("Sweep");
#pragma omp for private(j,ofs)
    for(i=1; i<size-1; ++i) {
      for(j=1; j<size-1; ++j) {
        ofs = i*size*size + j*size;
        jacobi_line(&phi[t1][ofs],&phi[t0][ofs],&phi[t0][ofs+size],&phi[t0][ofs-size],
			&phi[t0][ofs+size*size],&phi[t0][ofs-size*size],size);
/*
	for(k=1; k<size-1; ++k) {
	  phi[t1][ofs+k] = (  phi[t0][ofs+k-1]+phi[t0][ofs+k+1]
			      + phi[t0][ofs+k-size]+phi[t0][ofs+k+size]
			      + phi[t0][ofs+k-size*size]+phi[t0][ofs+k+size*size] )*oos;
	}
*/
      }
    }
    LIKWID_MARKER_STOP("Sweep");
}
    dummy(phi[0],phi[1]);
    t=t0; t0=t1; t1=t;
  }	  
  

  timing(&wct_end, &cput_end);
  runtime = wct_end-wct_start;
  iter *= 2;
  }

  iter /= 2;

  printf("size: %d  time: %lf  iter: %d  MLUP/s: %lf\n",size,runtime,iter,(double)iter*(size-2)*(size-2)*(size-2)/runtime/1000000.);

  LIKWID_MARKER_CLOSE;
  
  return 0;
}

File: julia/j3d/j3d.jl
using LIKWID
using Base.Threads
using Printf

const oos = 1.0/6.0

function jacobi_line(d, s, top, bottom, front, back, n)
    #pragma vector nontemporal
    for i in 2:n-1
        d[i] = oos*(s[i-1]+s[i+1]+top[i]+bottom[i]+front[i]+back[i])
    end
end

@noinline function dummy(a, b)
    return nothing
end

function main()
    Marker.init()

    length(ARGS) < 1 && error("Expected <size> as input argument!")

    size = parse(Int, ARGS[1])
    phi = [
        rand(size*size*size),
        rand(size*size*size)
    ]

    t1=1
    t2=2

    iter=1
    runtime=0.0
    while runtime<0.5
        # time measurement
        wct_start = time()
        for n in 1:iter
            @threads for i in 1:nthreads()
                Marker.startregion("Sweep")
            end
            @threads for i in 2:size-1
                for j in 2:size-1
                    ofs = (i-1)*size*size + (j-1)*size
                    @views jacobi_line(
                        phi[t2][ofs:end],
                        phi[t1][ofs:end],
                        phi[t1][ofs+size:end],
                        phi[t1][ofs-size:end],
			            phi[t1][ofs+size*size:end],
                        phi[t1][ofs-size*size:end],
                        size
                    )
                end
            end
            @threads for i in 1:nthreads()
                Marker.stopregion("Sweep")
            end
            dummy(phi[1], phi[2])
            t1, t2 = t2, t1
        end
        wct_end = time()
        runtime = wct_end - wct_start
        iter *= 2
    end

    iter /= 2

    @printf("size: %d  time: %lf  iter: %d  MLUP/s: %lf\n",size,runtime,iter,iter*(size-2)*(size-2)*(size-2)/runtime/1000000.0)

    Marker.close()
end

main()